com.unity.multiplayer.mlapi/Runtime/Spawning/NetworkPrefabHandler.cs

```diff
@@ -4,13 +4,36 @@
 namespace MLAPI.Spawning
 {
     /// <summary>
-    /// Interface for customizing asset spawn and destroy handlers
-    /// NOTE: Custom spawn and destroy handlers are only invoked on clients
+    /// Interface for customizing, overriding, spawning, and destroying Network Prefabs
+    /// Used by <see cref="NetworkPrefabHandler"/>
     /// </summary>
     public interface INetworkPrefabInstanceHandler
     {
-        NetworkObject HandleNetworkPrefabSpawn(ulong ownerClientId, Vector3 position, Quaternion rotation);
-        void HandleNetworkPrefabDestroy(NetworkObject networkObject);
+        /// <summary>
+        /// Client Side Only
+        /// Once an implementation is registered with the <see cref="NetworkPrefabHandler"/>, this method will be called every time
+        /// a Network Prefab associated <see cref="NetworkObject"/> is spawned on clients (excluding host since it is the server).
+        ///
+        /// Note on Pooling:  If you are using a NetworkObject pool, don't forget to make the NetworkObject active
+        /// via the  <see cref="GameObject.SetActive(bool)"/> method.
+        /// </summary>
+        /// <param name="ownerClientId">the owner for the <see cref="NetworkObject"/> to be instantiated</param>
+        /// <param name="position">the initial/default position for the <see cref="NetworkObject"/> to be instantiated</param>
+        /// <param name="rotation">the initial/default rotation for the <see cref="NetworkObject"/> to be instantiated</param>
+        /// <returns></returns>
+        NetworkObject ClientInstantiateOverride(ulong ownerClientId, Vector3 position, Quaternion rotation);
+
+        /// <summary>
+        /// Client and Server
+        /// Once an implementation is registered with the <see cref="NetworkPrefabHandler"/>, this method will be called every time
+        /// a Network Prefab associated <see cref="NetworkObject"/> is destroyed.
+        /// If <see cref="NetworkObject.Despawn(bool)"/> is invoked with the default destroy parameter (i.e. false) then this method will not be invoked.
+        ///
+        /// Note on Pooling: When invoked, you do not need to despawn or destroy the NetworkObject as long as you want your pool to persist.
+        /// The most common approach is to make the <see cref="NetworkObject"/> inactive by calling <see cref="GameObject.SetActive(bool)"/>.
+        /// </summary>
+        /// <param name="networkObject">The <see cref="NetworkObject"/> being destroyed</param>
+        void DestroyOverride(NetworkObject networkObject);
```

Comment:
I'm not sure if I'd go for `ClientInstantiateOverride` & `DestroyOverride` instead of `HandleNetworkPrefabSpawn` & `HandleNetworkPrefabDestroy` under an interface called `INetworkPrefabInstanceHandler`.

I think I'd drop "-Override" suffix and keep "Handle-" prefix around.

Also "Client-" prefix sounds a little risky to me in terms of our future potential and its API-breaking nature.
We might expand this feature's impact area and even make it work for both client and server sides. That would require another API-breaking change just to drop "Client-" prefix or we might have to add another super similar API that starts with "Server-" prefix to not break the old API (which would not be ideal IMHO).

Only thing I'd change here would be `HandleNetworkPrefabDestroy` → `HandleNetworkPrefabDespawn` to match `NetworkBehaviour.OnNetworkSpawn()` and `NetworkBehaviour.OnNetworkDespawn()` naming convention.

Also, it'd be tempting to keep client-specific comments around but I think they should go away and make feature non-client specific. Client-specific comment may go under `NetworkPrefabHandler` since it is the source where client-specific nature is coming from (and even that might change in the future to cover both client and server side, hence my earlier comments above).

Reply:
So the old naming here was very inconsistent. `Spawn`, `Destroy` don't match. We either want `HandleNetworkPrefabDespawn` or rename the `Spawn` to `Instantiate`.

Now the reason we want for `Instantiate/Destroy` was is because you can spawn/despawn a networkobject in MLAPI without triggering these handlers (GameObject.Destroy on the server for instance skips the handler). And what they really do is override a GameObject `Instatiate` / `Destroy` function and replace it with something else.

I prefer the `Override` suffix over `Handle` prefix. Because `HandleNetworkPrefabDestroy` does not indicate whether this handler is run in addition to the destroy or instead of the destroy. Dropping the `Client` prefix on `InstantiateOverride` sounds good if we consider future potential.

Reply:
so it's:

- `HandleNetworkPrefabSpawn`
- `HandleNetworkPrefabDespawn`

vs

- `InstantiateOverride`
- `DestroyOverride `

:)

I believe it's fair to say that we'd go for `Instantiate`/`Destroy`.

so my final suggestion would be:

- `HandleInstantiate`
- `HandleDestroy`

:P

Reply:
HandleInstantiate, HandleDestroy sounds good to me 👍

Reply:
The only issue I have with "HandleInstantiate" is that this is only invoked on the client side and could be misleading.  However, that can be made clear in documentation.

How about we even further simplify this?  I mean it is a "Handler" so not sure we have to put "handle" in front of all of the methods.  Why not:

- Instantiate

- Destroy

Reply:
> this is only invoked on the client side

hence my first comment, it is currently client-only but we could expand it to cover both server and client.

> it is a "Handler" so ...

I think that's also a fair point. `INetworkPrefabInstanceHandler.Instantiate` & `INetworkPrefabInstanceHandler.Destroy` it is :)